src/properties.ts

```diff
@@ -0,0 +1,347 @@
+      packageJsonParams['sonar.links.homepage'] = pkg.homepage;
+    }
+    if (pkg.bugs?.url) {
+      packageJsonParams['sonar.links.issue'] = pkg.bugs.url;
+    }
+    if (pkg.repository?.url) {
+      packageJsonParams['sonar.links.scm'] = pkg.repository.url;
+    }
+
+    const potentialCoverageDirs = [
+      // jest coverage output directory
+      // See: http://facebook.github.io/jest/docs/en/configuration.html#coveragedirectory-string
+      pkg['nyc']?.['report-dir'],
+      // nyc coverage output directory
+      // See: https://github.com/istanbuljs/nyc#configuring-nyc
+      pkg['jest']?.['coverageDirectory'],
+    ]
+      .filter(Boolean)
+      .concat(
+        // default coverage output directory
+        'coverage',
+      );
+    const uniqueCoverageDirs = Array.from(new Set(potentialCoverageDirs));
+    packageJsonParams['sonar.exclusions'] = sonarBaseExclusions;
+    for (const lcovReportDir of uniqueCoverageDirs) {
+      const lcovReportPath = path.posix.join(lcovReportDir, 'lcov.info');
+      if (fileExistsInProjectSync(lcovReportPath)) {
+        packageJsonParams['sonar.exclusions'] +=
+          (packageJsonParams['sonar.exclusions'].length > 0 ? ',' : '') +
+          path.posix.join(lcovReportDir, '**');
+        // https://docs.sonarsource.com/sonarqube/latest/analyzing-source-code/test-coverage/javascript-typescript-test-coverage/
+        packageJsonParams['sonar.javascript.lcov.reportPaths'] = lcovReportPath;
+        // TODO: use Generic Test Data to remove dependence of SonarJS, it is need transformation lcov to sonar generic coverage format
+      }
+    }
+
+    if (dependenceExists('mocha-sonarqube-reporter') && fileExistsInProjectSync('xunit.xml')) {
+      // https://docs.sonarqube.org/display/SONAR/Generic+Test+Data
+      packageJsonParams['sonar.testExecutionReportPaths'] = 'xunit.xml';
+    }
+    // TODO: (SCANNPM-13) use `glob` to lookup xunit format files and transformation to sonar generic report format
+  }
+  return packageJsonParams;
+}
+
+/**
+ * Convert CLI args into scanner properties.
+ */
+function getCommandLineProperties(cliArgs?: string[]): ScannerProperties {
+  if (!cliArgs || cliArgs.length === 0) {
+    return {};
+  }
+
+  // Parse CLI args (eg: -Dsonar.token=xxx)
+  const properties: ScannerProperties = {};
+  for (const arg of cliArgs) {
+    if (!arg.startsWith('-D')) {
+      continue;
+    }
+    const [key, value] = arg.substring(2).split('=');
+    properties[key] = value;
+  }
+
+  return properties;
+}
+
+/**
+ * Parse properties stored in sonar project properties file, if it exists.
+ */
+function getSonarFileProperties(projectBaseDir: string): ScannerProperties {
+  // Read sonar project properties file in project base dir
+  try {
+    const sonarPropertiesFile = path.join(projectBaseDir, SONAR_PROJECT_FILENAME);
+    const properties: ScannerProperties = {};
+    const data = fs.readFileSync(sonarPropertiesFile).toString();
+    const lines = data.split('\n');
```

Comment:
```suggestion
    const lines = data.split("(?:\r\n|\r|\n|\f)");
```

Reply:
Oh yes the Windows newline, thanks for spotting

Couldn't we just use `\r?\n|\f` though? I'd drop the non-capture flag `?:` and simplify `\r?\n|\r|\n` to `\r?\n`, if that sounds good
```suggestion
    const lines = data.split(/\r?\n|\f/);
```